src/DataViewer.jl

```diff
@@ -19,52 +22,76 @@ function viewdata(data::Array{Float64,1},t::AbstractArray{Float64,1}=linspace(0,
 	end
-	@materialize mouseposition, mouse_buttons_pressed,mouse_button_down, mouse_button_released,scroll = window.inputs
 
-	fscroll = droprepeats(map(Vec2f0, scroll))
-	pan = foldp(+, Vec2f0(0.0),  fscroll)
 
-	start_position = map(mouse_button_down) do button
-		value(mouseposition)
-	end
-	end_position = map(mouse_button_released) do button
-		ΔX = value(mouseposition)[1] - value(start_position)[1]
-		s = scalematrix(Vec3f0(1.0))
-		if ΔX > 0
-			s = translationmatrix(-Vec3f0(value(start_position)[1],0.0, 0.0))
-			t = scalematrix(Vec3f0(h/ΔX,1.0, 1.0))
-			s = s*t
-		else
-			push!(scroll, -value(pan)) #hackish way of reseting the pan signal
-		end
-		s
-	end
 
-	#TODO: This part doesn't really work
-	selection_rectangle = map(mouse_buttons_pressed) do button
-		ΔX = abs(value(mouseposition)[1] - value(start_position)[1])
-		ΔY = float(h)
-		if value(mouseposition)[1] > value(start_position)[1]
-			x = value(start_position)[1]
-		else
-			x = value(mouseposition)[1]
-		end
-		S = SimpleRectangle(x, 20.0, ΔX, ΔY)
-		S
-	end
-	#
 
-	new_model = map(pan,end_position) do _pan,_pos
-		t = translationmatrix(Vec3f0(_pan[1], 0.0, 0.0))
-		s = value(_pos)*t
-		s
-	end
+	_view(visualize(points, :lines, color=RGBA(0.0, 0.0, 0.0, 1.0)), window)
 
-	_view(visualize(points, :lines, color=RGBA(0.0, 0.0, 0.0, 1.0),model=new_model), window)
-	#TODO: This doesn't work
-	_view(visualize(value(selection_rectangle), model=end_position), window)
 	#
 	renderloop(window)
 end
 
+
+# it's time to have these defined in GeometryTypes
+function Rect(x, y, w, h)
+    SimpleRectangle(round(Int, x), round(Int, y), round(Int, w), round(Int, h))
+end
+function Rect(xy::FixedVector, w, h)
+    Rect(xy[1], xy[2], w, h)
+end
+function Rect(x, y, wh::FixedVector)
+    Rect(x, y, wh[1], wh[2])
+end
+function Rect(xy::FixedVector, wh::FixedVector)
+    Rect(xy[1], xy[2], wh[1], wh[2])
+end
+
+function register_drag_zoom(cam, screen, key = GLFW.MOUSE_BUTTON_LEFT)
+    @materialize mouseposition, mouse_buttons_pressed = screen.inputs
+    @materialize mouse_button_down, mouse_button_released = screen.inputs
+
+    is_dragging = false
+    rect = Rect(0,0,0,0)
+    dragged_rect = foldp(
+            (is_dragging, rect),
+            mouse_buttons_pressed, mouseposition
+        ) do v0, m_pressed, m_pos
+        was_dragging, rect = v0
+        keypressed = (length(m_pressed) == 1) && (key in m_pressed)
+        p = imagespace(m_pos, cam)
+        if was_dragging
+            wh = p - minimum(rect)
+            rect = Rect(minimum(rect), wh)
+            if keypressed # was dragging and still dragging
+                return true, rect
+            else
+                return false, rect # anything else will stop the dragging
+            end
+        elseif keypressed # was not dragging, but now key is pressed
+            return true, Rect(p, 0, 0)
+        end
+        return v0
+    end
+    lw = 2f0
+    rect_vis = visualize(
```

Comment:
I tried incorporating these changes into my code, but this line gave the following error:
ERROR: can't splice Array{Float32,1} into an OpenGL shader. Make sure all fields are of a concrete type and isbits(FieldType)-->true
 in toglsltype_string(::Array{Float32,1}) at /Users/roger/.julia/v0.5/GLAbstraction/src/GLUniforms.jl:101
 in mustache2replacement(::SubString{String}, ::Dict{String,String}, ::Dict{Symbol,Any}) at /Users/roger/.julia/v0.5/GLAbstraction/src/GLShader.jl:313
 in #75 at /Users/roger/.julia/v0.5/GLAbstraction/src/GLShader.jl:328 [inlined]
 in insert_from_view at /Users/roger/.julia/v0.5/GLAbstraction/src/GLShader.jl:243 [inlined]
 in mustache_replace(::GLAbstraction.##75#76{Dict{String,String},Dict{Symbol,Any},Dict{String,String}}, ::String) at /Users/roger/.julia/v0.5/GLAbstraction/src/GLShader.jl:275
 in template2source(::String, ::Dict{String,String}, ::Dict{Symbol,Any}) at /Users/roger/.julia/v0.5/GLAbstraction/src/GLShader.jl:327
 in (::GLAbstraction.##65#66{String,Dict{String,String},Dict{Symbol,Any}})() at /Users/roger/.julia/v0.5/GLAbstraction/src/GLShader.jl:165
 in get!(::GLAbstraction.##65#66{String,Dict{String,String},Dict{Symbol,Any}}, ::Dict{String,Array{String,1}}, ::String) at ./dict.jl:663
 in gl_convert(::GLVisualize.GLVisualizeShader, ::Dict{Symbol,Any}) at /Users/roger/.julia/v0.5/GLAbstraction/src/GLShader.jl:224
 in GLAbstraction.RenderObject{Pre}(::Dict{Symbol,Any}, ::GLVisualize.GLVisualizeShader, ::GLAbstraction.StandardPrerender, ::Void, ::GeometryTypes.HyperRectangle{3,Float32}, ::Void) at /Users/roger/.julia/v0.5/GLAbstraction/src/GLTypes.jl:295
 in assemble_robj(::Dict{Symbol,Any}, ::GLVisualize.GLVisualizeShader, ::GeometryTypes.HyperRectangle{3,Float32}, ::UInt32, ::Void, ::Void) at /Users/roger/.julia/v0.5/GLVisualize/src/utils.jl:32
 in assemble_shader(::Dict{Symbol,Any}) at /Users/roger/.julia/v0.5/GLVisualize/src/utils.jl:59
 in visualize(::Any, ::GLAbstraction.Style{:lines}, ::Dict{Symbol,Any}) at /Users/roger/.julia/v0.5/GLVisualize/src/visualize_interface.jl:21
 in (::GLVisualize.#kw##visualize)(::Array{Any,1}, ::GLVisualize.#visualize, ::Any, ::Symbol) at ./<missing>:0